samples/snake/src/main/kotlin/example/SnakeViewModel.kt

```diff
@@ -0,0 +1,286 @@
+package example
+
+import example.models.Direction
+import example.models.Food
+import example.models.GameState
+import example.models.INVALID_FOOD
+import example.models.Phase
+import example.models.Point
+import example.models.Size
+import example.models.Snake
+import example.models.copy
+import example.models.head
+import example.models.height
+import example.models.rawHead
+import example.models.width
+import example.models.x
+import example.models.y
+import example.utils.asLong
+import java.util.concurrent.atomic.AtomicReference
+import kotlin.random.Random
+import kotlin.system.exitProcess
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.distinctUntilChanged
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.flow.onEach
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+private const val INITIAL_SNAKE_SIZE = 3
+
+class SnakeViewModel(
+	private val fieldSize: Size = Size(60, 20),
+	private val updateDelayMs: Long = 200L,
+) {
+	private val coroutineScope = CoroutineScope(Dispatchers.Default)
+
+	private val gameState = AtomicReference(GameState(snake = createSnake()))
+
+	private val updateUiStateFlow = MutableSharedFlow<Unit>(
+		extraBufferCapacity = 1,
+		onBufferOverflow = BufferOverflow.DROP_OLDEST,
+	)
+
+	private val moveDirectionFlow = MutableSharedFlow<Direction>(
+		extraBufferCapacity = 1,
+		onBufferOverflow = BufferOverflow.DROP_LATEST,
+	)
+
+	val uiStateFlow: StateFlow<SnakeUiState> = updateUiStateFlow
+		.map { gameState.get() }
+		.onEach { checkGameOver(it) }
+		.map { it.toUiState() }
+		.distinctUntilChanged()
+		.stateIn(coroutineScope, SharingStarted.Lazily, gameState.get().toUiState())
```

Comment:
In general this isn't my favorite state management. It's a bit of a state soup. I would think we can put a bit more directly into composables rather than pulling it completely out. Can deal with it later.